File: irrgartenjaava/src/irrgartenjaava/LabyrinthCharacter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgartenjaava;

/**
 *
 * @author jxlig0d
 * 
 * @brief P4, clase abstracta de las entidades del laberinto
 * 
 */
abstract class LabyrinthCharacter {
    
    // Variable que guarda el name
    private String name;
    
    // Variable que guarda la inteligencia
    private float intelligence;
    
    // Variable que guarda la fuerza
    private float strength;
    
    // Variable que guarda la vida
    private float health;
    
    // Variable que guarda el row
    private int row;
    
    // Variable que guarda el col
    private int col;
    
    private static final int INVALIDA=-1;
    
    
    // constructor por defecto.
    public LabyrinthCharacter(String name, float intelligence, float strength, float health){
        
        this.name=name;
        this.intelligence=intelligence;
        this.strength=strength;
        this.health=health;
        
        this.row=INVALIDA;
        this.col=INVALIDA;
    }
    
    //constructor de copia.
    public LabyrinthCharacter(LabyrinthCharacter other){
        
        this.name=other.name;
        this.intelligence=other.intelligence;
        this.strength=other.strength;
        this.health=other.health;
        
        this.row=other.row;
        this.col=other.col;
    }
    
    
    //----------------------------------------------------------------------------------------------------
    // getters y setters y gotWounded TODOS PROTECTED
    
    protected float getIntelligence(){
        return this.intelligence;
    }
    
    protected float getStrength(){
        return this.strength;
    }
    
    protected float getHealth(){
        return this.health;
    }
    
    protected void setHealth(float health){
        this.health=health;
    }
    
    protected void gotWounded(){
        this.health=this.health-1;
    }
    
    
    //----------------------------------------------------------------------------------------------------
    // RESTO DE FUNCIONES PUBLICAS
    
    public boolean dead(){
        return (this.health<=0);
    }
    
    public int getRow(){
        return this.row;
    }
    
    public int getCol(){
        return this.col;
    }
    
    public void setPos(int row, int col){
        this.col=col;
        this.row=row;
    }
    
    @Override
    public String toString(){
         
        //en formatom
        final String FORMAT = "%.6f";
        String toReturn=this.name+"["+"intelligence:"+ String.format(FORMAT, this.intelligence) + ", strength:"+String.format(FORMAT, this.strength);
        toReturn+=", health:"+String.format(FORMAT, this.health)+", pos:("+this.row+", "+this.col+")]";
        
        return toReturn;
    }
    
    
    // Estos abstractos no hace falta rellenarlos
    // funcionan como una especie de .h de c++
   
    public abstract float attack();
    
    public abstract boolean defend(float attack);
    
}

File: irrgarten_ruby/src/labyrinth_character.rb
#encoding:utf-8

# Abstract no se declara directamente aqui como en JAVA
module Irrgarten

  class LabyrinthCharacter

    @@INVALID_POS = -1

    def initialize(name, intelligence, strength, health)
      @name = name.to_s
      @intelligence = intelligence.to_f
      @strength = strength.to_f
      @health = health.to_f
      
      @row = @@INVALID_POS
      @col = @@INVALID_POS

    end  

    # consultores protected

    protected
    attr_reader :name, :intelligence, :strength
    attr_accessor :health

    public
    attr_reader :row, :col

    # metodos publicos
    
    def pos(row, col)
      @row = row  
      @col = col
    end

    def copy(other)
      @name = other.name
      @intelligence = other.intelligence
      @strength = other.strength
      @health = other.health
      pos(other.row, other.col)
    end
    
    def dead
      return @health <= 0
    end

    protected

    def got_wounded()
      @health -= 1
    end

    public

    def to_s

      return "#{@name} (#{@intelligence}, #{@strength}, #{@health}, #{@row}, #{@col})"

    end

    # ESTAS CLASES SE IMPLEMENTAN EN LAS CLASES HIJAS
    # asi se hace la herencia aqui xd??
    
    def attack
      raise NotImplementedError
    end

    def defend
      raise NotImplementedError
    end


  end #class
end #module